vue/sbc-common-components/src/services/token.services.ts

```diff
@@ -0,0 +1,119 @@
+import Keycloak, { KeycloakInitOptions, KeycloakInstance } from 'keycloak-js'
+
+class TokenServices {
+  private kc: KeycloakInstance
+  private static instance: TokenServices
+  private kcLoaded
+  private counter = 0
+  private timerId
+  private kcOptions: KeycloakInitOptions = {
+    onLoad: 'login-required',
+    checkLoginIframe: false,
+    timeSkew: 0,
+    token: sessionStorage.getItem('KEYCLOAK_TOKEN'),
+    refreshToken: sessionStorage.getItem('KEYCLOAK_REFRESH_TOKEN'),
+    idToken: sessionStorage.getItem('KEYCLOAK_ID_TOKEN')
+  }
+
+  initUsingKc (kcInstance: KeycloakInstance) {
+    console.info('[TokenServices] KC Instance registered using kc instance')
+    this.kc = kcInstance
+  }
+
+  initUsingUrl (keyCloakConfigurl: string) {
+    console.info('[TokenServices] KC Instance registered using kc URL')
+    var self = this
+    return new Promise((resolve, reject) => {
+      this.kc = Keycloak(keyCloakConfigurl)
+      this.kc.init(this.kcOptions
+      ).success(function (authenticated) {
+        console.info('[TokenServices] is User Authenticated:Syncing' + authenticated)
+        self.syncSessionStorage()
+        resolve(self.kc.token)
+      }).error(function (err) {
+        console.info('[TokenServices] Fatal Error:Cudnt Initialise KC instance' + err)
+        reject(new Error('Cudnt Initialise KC'))
+      })
+    })
+  }
+
+  scheduleRefreshTimer () {
+    console.info('[TokenServices Starting the timer] ')
+    this.scheduleRefreshToken()
+  }
+
+  refreshToken () {
+    console.log('[TokenServices] One time Token Refreshing ')
+    return new Promise((resolve, reject) => {
+      this.kc.updateToken(-1)
+        .success(refreshed => {
+          if (refreshed) {
+            console.log('[TokenServices] One time Token Refreshed ')
+            this.syncSessionStorage()
+            resolve()
+          }
+        })
+        .error(() => {
+          // this.cleanupSession()
+          reject(new Error('Cudnt refresh Token'))
+        })
+    })
+  }
+
+  stopRefreshTimer () {
+    console.info('[TokenServices Stopping the timer] ')
+    clearTimeout(this.timerId)
+  }
+
+  private scheduleRefreshToken () {
+    let self = this
+
+    // check if refresh token is still valid . Or else clear all timers and throw errors
+    let refreshTokenExpiresIn = self.kc.refreshToken['exp'] - Math.ceil(new Date().getTime() / 1000) + self.kc.timeSkew
+    if (refreshTokenExpiresIn < 0) {
+      throw new Error('Refresh Token Expired..No more token refreshes')
+    }
+
+    let expiresIn = self.kc.tokenParsed['exp'] - Math.ceil(new Date().getTime() / 1000) + self.kc.timeSkew
+    let refreshInMilliSeconds = (expiresIn * 1000) - 2000 // refresh before two second of expiry
```

Comment:
2 seconds seems very short... it doesn't give much time to refresh the token. The network may be congested, and the in-browser timer may slow down for a lot of reasons:
- the CPU is overloaded
- the browser tab is in the background mode
- the laptop is on battery

Is there a recommended value for this? 30 seconds maybe?

Ref: https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Reasons_for_delays_longer_than_specified

Reply:
Will add it as a config may be... 
30 seconds is too much..  this is the time  before expiry when we are refreshing..so 5 seconds or 10 seconds is good enough

Reply:
Another reason why 2 seconds is probably too short:

![skew](https://user-images.githubusercontent.com/10002889/70758884-3e8bb880-1cf9-11ea-862a-1ff5c82644b0.PNG)

Reply:
Yup..5 0r 10 is good..